zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -2921,10 +2988,47 @@ private void addChildListener(String path, IZkChildListener listener) {
     listeners.add(listener);
   }
 
+  private void addPersistListener(String path, Object listener) {
+    try {
+      _persistListenerMutex.lockInterruptibly();
+      if (listener instanceof IZkChildListener) {
+        addChildListener(path, (IZkChildListener) listener);
+      } else if (listener instanceof IZkDataListener) {
+        addDataListener(path, (IZkDataListener) listener);
+      }
+    } catch (InterruptedException ex) {
+      throw new ZkInterruptedException(ex);
+    } finally {
+      _persistListenerMutex.unlock();
```

Comment:
Would it be possible to make these pieces of code generic?

_mutx.lock();
passedin method.invoke(xxxx)

catch(e){}
finally{
_mutex.unlock();
}

Reply:
TFTR. Good point. Will update.